src/Common/src/System/IO/PathInternal.cs

```diff
@@ -23,5 +25,67 @@ internal static void CheckInvalidPathChars(string path, bool checkAdditional = f
             if (PathInternal.HasIllegalCharacters(path, checkAdditional))
                 throw new ArgumentException(SR.Argument_InvalidPathChars, "path");
         }
+
+
+        /// <summary>
+        /// Returns true if the given StringBuilder starts with the given value.
+        /// </summary>
+        /// <param name="value">The string to compare against the start of the StringBuilder.</param>
+        internal static bool StartsWithOrdinal(this StringBuilder builder, string value)
+        {
+            if (value == null || builder.Length < value.Length)
+                return false;
+
+            for (int i = 0; i < value.Length; i++)
+            {
+                if (builder[i] != value[i]) return false;
+            }
+            return true;
+        }
+
+        /// <summary>
+        /// Returns true if the given string starts with the given value.
+        /// </summary>
+        /// <param name="value">The string to compare against the start of the source string.</param>
+        internal static bool StartsWithOrdinal(this string source, string value)
+        {
+            if (value == null || source.Length < value.Length)
+                return false;
+
+            for (int i = 0; i < value.Length; i++)
+            {
+                if (source[i] != value[i]) return false;
+            }
+            return true;
+        }
+
+        /// <summary>
+        /// Trims the specified characters from the end of the StringBuilder.
+        /// </summary>
+        internal static StringBuilder TrimEnd(this StringBuilder builder, params char[] trimChars)
+        {
+            if (trimChars == null || trimChars.Length == 0)
+                return builder;
+
+            int end = builder.Length - 1;
+
+            for (; end >= 0; end--)
+            {
+                int i = 0;
+                char ch = builder[end];
```

Comment:
This is probably completely fine, but at some point it'd be interesting to do some profiling with long paths to see if this shows up at all.  StringBuilder is now implemented as a linked list of builders, so indexing into it actually needs to walk that list to find the desired index.